pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,175 @@
+package namespace
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/rand"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog"
+
+	"github.com/kcp-dev/kcp/pkg/apis/cluster/v1alpha1"
+)
+
+const clusterLabel = "kcp.dev/cluster"
+
+// reconcileResource is responsible for setting the cluster for a resource of
+// any type, to match the cluster where it's namespace is assigned.
+func (c *Controller) reconcileResource(ctx context.Context, unstr *unstructured.Unstructured, gvr *schema.GroupVersionResource) error {
+	klog.Infof("reconciling %s %s/%s", gvr.String(), unstr.GetNamespace(), unstr.GetName())
+
+	lbls := unstr.GetLabels()
+	if lbls == nil {
+		lbls = map[string]string{}
+	}
+
+	// If the resource is not namespaced (incl if the resource is itself a
+	// namespace), ignore it.
+	if unstr.GetNamespace() == "" {
+		klog.Infof("%s %s had no namespace; ignoring", gvr.String(), unstr.GetName())
+		return nil
+	}
+
+	// Align the resource's assigned cluster with the namespace's assigned
+	// cluster.
+	// First, get the namespace object (from the cached lister).
+	ns, err := c.namespaceLister.Get(unstr.GetNamespace())
+	if err != nil {
+		return err
+	}
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+
+	if lbls[clusterLabel] == ns.Labels[clusterLabel] {
+		// Already assigned to the right cluster.
+		return nil
+	}
+
+	// Update the resource's assignment. This will be applied when
+	// reconcile is done.
+	klog.Infof("updating %s %s/%s cluster %q -> %q", gvr.String(), unstr.GetNamespace(), unstr.GetName(), lbls[clusterLabel], ns.Labels[clusterLabel])
+	lbls[clusterLabel] = ns.Labels[clusterLabel]
```

Comment:
What if the namespace doesn't have the clusterLabel set?

Reply:
Then the resources will also have their clusterLabel unset, if they weren't already. I believe this is desired behavior, since it means a resource belonging to an unassigned namespace can be ensured to also be unassigned, in case something outside of this code set the label for some reason.